algebra/vector_space.lean

```diff
@@ -0,0 +1,305 @@
+by {unfold ker at *, simp [HV]}
+
+instance : subspace K V A.ker :=
+{
+    p_add := ker_add,
+    p_zero := ker_zero,
+    p_neg := ker_neg,
+    p_smul := ker_smul
+}
+
+end ker
+
+
+section Hom
+
+variables {u v : V} {A B C : linear_space K V W}
+
+def add (A B : linear_space K V W) : (linear_space K V W) :=
+{
+    T       := (λ v, (A.T v) + (B.T v)),
+    map_add := (λ u v, by simp),
+    map_mul := (λ c v, by simp [smul_left_distrib])
+}
+
+def zero : linear_space K V W :=
+{
+    T       := (λ v, 0),
+    map_add := (λ u v, by simp),
+    map_mul := (λ c v, by simp)
+}
+
+def neg (A : linear_space K V W) : linear_space K V W :=
+{
+    T       := (λ v, -A.T v),
+    map_add := (λ u v, by simp),
+    map_mul := (λ c v, by simp)
+}
+
+instance : has_add (linear_space K V W) := ⟨add⟩
+instance : has_zero (linear_space K V W) := ⟨zero⟩
+instance : has_neg (linear_space K V W) := ⟨neg⟩
+
+@[simp] lemma add_simp : (A + B).T v = A.T v + B.T v := rfl
+@[simp] lemma zero_simp : (0:linear_space K V W).T v = 0 := rfl
+@[simp] lemma neg_simp : (-A).T v = -(A.T v) := rfl
+
+def smul (c : K) (A : linear_space K V W) : linear_space K V W :=
+{
+    T       := (λ v, c•(A.T v)),
+    map_add := (λ u v, by simp [smul_left_distrib]),
+    map_mul := (λ k v, by simp [smul_smul])
+}
+
+local infix `⬝` := smul
+
+@[simp] lemma smul_simp : (c⬝A).T v = c•(A.T v) := rfl
+
+variables (c d u v A B C)
+
+theorem add_zero : A + 0 = A := ext (λ v, by simp [add_zero])
+theorem zero_add : 0 + A = A := ext (λ v, by simp [zero_add])
+theorem add_right_neg : A + (-A) = 0 := ext (λ v, by simp [add_right_neg])
+theorem add_left_neg : (-A) + A = 0 := ext (λ v, by simp [add_left_neg])
+theorem add_assoc : A + B + C = A + (B + C) := ext (λ v, by simp [add_assoc])
+theorem add_comm : A + B = B + A := ext (λ v, by simp [add_comm])
+theorem smul_left_distrib : c⬝(A+B) = c⬝A + c⬝B := ext (λ v, by simp [smul_left_distrib])
+theorem smul_right_distrib : (c+d)⬝A = c⬝A + d⬝A := ext (λ v, by simp [smul_right_distrib])
+theorem mul_smul : (c*d)⬝A = c⬝(d⬝A) := ext (λ v, by simp [mul_smul])
+theorem one_smul : 1⬝A = A := ext (λ v, by simp [one_smul])
+
+instance : vector_space K (linear_space K V W) :=
+{
+    add                 := add,
+    add_assoc           := add_assoc,
+    zero                := zero,
+    zero_add            := zero_add,
+    add_zero            := add_zero,
+    neg                 := neg,
+    add_left_neg        := add_left_neg,
+    add_comm            := add_comm,
+    smul                := smul,
+    smul_left_distrib   := smul_left_distrib,
+    smul_right_distrib  := smul_right_distrib,
+    mul_smul            := mul_smul,
+    one_smul            := one_smul
+}
+
+end Hom
+
+def Hom {K : Type} (V W : Type) [field K] [vector_space K V] [vector_space K W] :=
+vector_space K (linear_space K V W)
+
+def vector_space.dual {K: Type} (V: Type) [field K] [vector_space K V] :=
+@Hom K V K _ _ field.to_vector_space
+
+-- \ast
+postfix `∗`:100 := vector_space.dual
```

Comment:
Don't define a global notation unless you use it. Someone is bound to want to use this character for something else later, and it will cause overloading problems.